namespace/http_resolver.go

```diff
@@ -0,0 +1,423 @@
+		case html.ErrorToken:
+			if z.Err() == io.EOF {
+				break ParsingLoop
+			}
+			return nil, z.Err()
+		case html.StartTagToken, html.SelfClosingTagToken:
+			switch tagName, hasAttr := z.TagName(); string(tagName) {
+			case "head":
+				if readingMetaTags {
+					return nil, fmt.Errorf("unexpected head tag")
+				}
+			case "meta":
+				if !hasAttr {
+					return nil, fmt.Errorf("meta tag empty")
+				}
+				scp, newEntries, err := parseHTMLMetaTag(z, name)
+				if err != nil {
+					return nil, err
+				}
+				if scp == "" {
+					for _, entry := range newEntries {
+						if err = entries.Add(entry); err != nil {
+							return nil, err
+						}
+					}
+				} else if parsedScope != "" {
+					return nil, fmt.Errorf("multiple scopes defined")
+				} else {
+					parsedScope = scp
+				}
+			default:
+				return nil, fmt.Errorf("unexpected html element %q", string(tagName))
+			}
+			readingMetaTags = true
+		case html.EndTagToken:
+			switch tagName, _ := z.TagName(); string(tagName) {
+			case "head":
+				break ParsingLoop
+			case "meta":
+			default:
+				return nil, fmt.Errorf("unexpected tag %q", tagName)
+			}
+		default:
+			continue ParsingLoop
+		}
+	}
+	if !readingMetaTags || len(entries.entries) == 0 {
+		return nil, fmt.Errorf("no entries found")
+	}
+	if parsedScope == "" {
+		parsedScope, err = parseScope(name)
+		if err != nil {
+			return nil, fmt.Errorf("cannot use given name %q as scope: %v", name, err)
+		}
+	}
+	for i := range entries.entries {
+		entries.entries[i].scope = parsedScope
+	}
+	return entries, nil
+}
+
+type httpResolver struct {
+	config *HTTPResolverConfig
+	cache  map[string]cacheEntry
+}
+
+/* Create base HTTP resolver.
+ *
+ * It uses discovery process to fetch scope entries. Multiple discovery
+ * endpoints may be queried during single `Resolve()` call depending on scope's
+ * namespace extensions and given callback.
+ *
+ * It caches fetched entries for further lookups.
+ */
+func NewHTTPResolver(config *HTTPResolverConfig) Resolver {
+	if config == nil {
+		config = &HTTPResolverConfig{}
+	}
+	if config.Client == nil {
+		config.Client = &http.Client{}
+	}
+	if config.ResolverFactory == nil {
+		config.ResolverFactory = func(entries *Entries) Resolver {
+			return NewSimpleResolver(entries, true)
+		}
+	}
+	if config.NSResolveCallback == nil {
+		config.NSResolveCallback = func(name string, namespace scope) NSResolveActionEnum {
+			if !namespace.Contains(name) {
+				logrus.Debugf("Ignoring extension namespace %q which isn't an ancestor of %q", namespace, name)
+				return NSResolveActionIgnore
+			}
+			return NSResolveActionRecurse
+		}
+	}
+	return &httpResolver{
+		config: config,
+		cache:  make(map[string]cacheEntry),
+	}
+}
+
+func (hr *httpResolver) nameToURL(name string) string {
+	return "https://" + name + "?docker-discovery=1"
+}
+
+func (hr *httpResolver) resolveEntries(es *Entries, visited map[string]struct{}, name string) error {
+	cached, exists := hr.cache[name]
```

Comment:
On the second thought, this will prevent `httpResolver` from reusing cached entries during recursion. If we have:

    multiResolver{
        resolvers: []Resolver {
            cacheResolver{ resolver: &httpResolver{...} },
            cacheResolver{ resolver: &httpResolver{...} },
            cacheResolver{ resolver: &httpResolver{...} },
            ...
        }
    }

there are separate caches. So even though the namespace entry being resolved at the 3rd level of recursion is stored in the 1st level cache, it will be fetched again.

Now I've got an idea. What about shared cache?

    func NewCacheResolver(baseResolver Resolver, cache *SharedCache) (Resolver, SharedCache) {
        if cache == nil {
            cache := newSharedCache()
        }
        ...
        return resolver, cache
    }

    ...

    r, cache := NewCacheResolver(NewHttpResolver(...), nil)
    NewMultiResolver(r, NewCacheResolver(NewHttpResolver(...), cache), ...)

Hmm, this bring me to a solution without any shared cache:

    multiResolver{
        resolvers: []Resolver {
            // the same cache resolver all over again
            cacheResolver, cacheResolver, cacheResolver, ...
        }
    }

But I think the concept of shared cache would be useful. I'll give it a try. Please stop me if you have better idea.